sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java

```diff
@@ -66,48 +99,105 @@ public FileChecksumMatcher(String checksum, String filePath) {
 
     this.expectedChecksum = checksum;
     this.filePath = filePath;
+    this.shardTemplate =
+        Pattern.compile(shardTemplate == null ? DEFAULT_SHARD_TEMPLATE : shardTemplate);
   }
 
   @Override
   public boolean matchesSafely(PipelineResult pipelineResult) {
+    // Load output data
+    List<String> outputs;
     try {
-      // Load output data
-      List<String> outputs = readLines(filePath);
+      outputs = readFilesWithRetries(Sleeper.DEFAULT, BACK_OFF_FACTORY.backoff());
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Failed to read from: %s", filePath), e);
+    }
 
-      // Verify outputs. Checksum is computed using SHA-1 algorithm
-      actualChecksum = hashing(outputs);
-      LOG.info("Generated checksum for output data: {}", actualChecksum);
+    // Verify outputs. Checksum is computed using SHA-1 algorithm
+    actualChecksum = computeHash(outputs);
+    LOG.info("Generated checksum: {}", actualChecksum);
```

Comment:
Silence this or `LOG.debug`. If there is a mismatch the error thrown should contain this information.